File: android/app/src/main/java/ae/findmybay/data/api/BookingApi.kt
package ae.findmybay.data.api

import ae.findmybay.data.api.dto.BookingDto
import ae.findmybay.data.api.dto.BookingListResponse
import ae.findmybay.data.api.dto.BookingQrDto
import ae.findmybay.data.api.dto.CancelBookingResponse
import ae.findmybay.data.api.dto.CreateBookingBody
import ae.findmybay.data.api.dto.ReviewBody
import ae.findmybay.data.api.dto.ReviewDto
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.Path

interface BookingApi {

    @POST("bookings")
    suspend fun create(@Body body: CreateBookingBody): BookingDto

    @GET("bookings/me")
    suspend fun mine(): BookingListResponse

    @GET("bookings/{id}/qr")
    suspend fun qr(@Path("id") id: String): BookingQrDto

    @POST("bookings/{id}/cancel")
    suspend fun cancel(@Path("id") id: String): CancelBookingResponse

    @POST("bookings/{id}/review")
    suspend fun submitReview(
        @Path("id") id: String,
        @Body body: ReviewBody,
    ): ReviewDto

    @GET("bookings/{id}/review")
    suspend fun getReview(@Path("id") id: String): ReviewDto
}

File: ios/FindMyBay/Data/API/BookingAPI.swift
// BookingAPI.swift
//
// Booking endpoints — parity with `data/api/BookingApi.kt`.

import Foundation

protocol BookingAPI {
    func create(_ body: CreateBookingBody) async throws -> BookingDTO
    func mine() async throws -> BookingListResponse
    func qr(id: String) async throws -> BookingQrDTO
    func cancel(id: String) async throws -> CancelBookingResponse
    func submitReview(id: String, body: ReviewBody) async throws -> ReviewDTO
    func getReview(id: String) async throws -> ReviewDTO
}

struct BookingAPIImpl: BookingAPI {
    let client: APIClient

    func create(_ body: CreateBookingBody) async throws -> BookingDTO {
        try await client.request(.POST, path: "bookings", body: body)
    }

    func mine() async throws -> BookingListResponse {
        try await client.request(.GET, path: "bookings/me")
    }

    func qr(id: String) async throws -> BookingQrDTO {
        try await client.request(.GET, path: "bookings/\(id)/qr")
    }

    func cancel(id: String) async throws -> CancelBookingResponse {
        try await client.request(.POST, path: "bookings/\(id)/cancel")
    }

    func submitReview(id: String, body: ReviewBody) async throws -> ReviewDTO {
        try await client.request(.POST, path: "bookings/\(id)/review", body: body)
    }

    func getReview(id: String) async throws -> ReviewDTO {
        try await client.request(.GET, path: "bookings/\(id)/review")
    }
}
